### PR TITLE
Run work posted to test synchronization context for up to ten seconds after the test method completes

### LINKED ISSUE #3209: Test fail when posting to SynchronizationContext.Current
# Issue
Our code is relying on a third-party library which posts some action to be executed on the `SynchronizationContext.Current`, this breaks some tests that require `Apartment(ApartmentState.STA)` and return `async Task`.
We managed to strip down the problem to the following tests (one is failing while the other is not).

Tested with NUnit console 3.10.0 and NUnit framework 3.11.0

## Failing test
```C#
[Test]
[Apartment(ApartmentState.STA)]
public async Task FailingTest()
{
    SynchronizationContext.Current.Post(state => { }, null);

    await Task.CompletedTask;
}
```
❗️  Exception
```
System.InvalidOperationException : Shutting down SingleThreadedTestSynchronizationContext with work still in the queue.
   in NUnit.Framework.Internal.SingleThreadedTestSynchronizationContext.ShutDown() in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\SingleThreadedSynchronizationContext.cs:riga 90
   in NUnit.Framework.Internal.AsyncToSyncAdapter.<>c__DisplayClass10_1.<InitializeExecutionEnvironment>b__0() in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncToSyncAdapter.cs:riga 128
   in NUnit.Framework.Internal.AsyncToSyncAdapter.Await(Func`1 invoke) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\AsyncToSyncAdapter.cs:riga 88
   in NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context) in C:\src\nunit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs:riga 84
```

## Passing test
```C#
[Test]
[Apartment(ApartmentState.STA)]
public async Task SuccessfulTest()
{
    SynchronizationContext.Current.Post(state => { }, null);

    await Task.Yield();
}
```

## Comments
We understood that `Task.Yield()` will force a context switch allowing for the synchronization context queue to be processed, while `Task.Result` won't. After having a look at the code and issues we saw that this is linked to #2818 and #2774

### PR DESCRIPTION
**Pull request by jnm2**
Fixes #3209.

### CODE REVIEW

File: src/NUnitFramework/tests/TestUtils.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+using System;
+using System.Threading;
+using NUnit.Framework.Internal;
+
+namespace NUnit.Framework
```
Comment (rprouse) [inline src/NUnitFramework/tests/TestUtils.cs]: Should this class be in the folder/namespace `TestUtilities` like the previous class?

File: src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs [hunk 2]
```diff
@@ -87,35 +109,37 @@ public void ShutDown()
...
             lock (_queue)
             {
-                status = Status.ShutDown;
-                Monitor.Pulse(_queue);
+                switch (_status)
+                {
+                    case Status.ShuttingDown:
+                    case Status.ShutDown:
+                        break;
+                }
```
[the remark this answers is not in the document]
Comment (jnm2) [reply src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs]: Good question, haha! It's supposed to be `return;`. I'll write a test enforcing this.

File: src/NUnitFramework/tests/TestUtils.cs (hunk 1 above)
Comment (jnm2) [reply src/NUnitFramework/tests/TestUtils.cs]: Sure, was 50/50 on this.

Comment (jnm2) [reply src/NUnitFramework/tests/TestUtils.cs]: Ugh, also forgot the license boilerplate.

**Comment by jnm2** [comment]
@rprouse I added a new commit in which I changed the no-op switch to `return;` and added a new test demonstrating the need. I rebased the other change in. Ready for review.

**Pull request merged**